prompb/io/prometheus/client/decoder.go

```diff
@@ -0,0 +1,775 @@
+	}
+	m.resetMetricFamily()
+	m.mfData = b[varIntLength:totalLength] // TODO: Copy and reuse slice.
+
+	m.inPos += totalLength
+	return m.MetricFamily.unmarshalWithoutMetrics(m, m.mfData)
+}
+
+func (m *MetricStreamingDecoder) resetMetricFamily() {
+	m.metrics = m.metrics[:0]
+	m.metricIndex = 0
+	m.MetricFamily.Reset()
+}
+
+func (m *MetricStreamingDecoder) NextMetric() error {
+	if m.metricIndex >= len(m.metrics) {
+		return io.EOF
+	}
+
+	m.resetMetric()
+	m.mData = m.mfData[m.metrics[m.metricIndex].start:m.metrics[m.metricIndex].end] // TODO: Copy and reuse slice.
+	if err := m.Metric.unmarshalWithoutLabels(m, m.mData); err != nil {
+		return err
+	}
+	m.metricIndex++
+	return nil
+}
+
+func (m *MetricStreamingDecoder) resetMetric() {
+	m.labels = m.labels[:0]
+	m.TimestampMs = 0
+
+	// TODO(bwplotka): Autogenerate reset functions.
+	if m.Metric.Counter != nil {
+		m.Metric.Counter.Value = 0
+		m.Metric.Counter.CreatedTimestamp = nil
+		m.Metric.Counter.Exemplar = nil
+	}
+	if m.Metric.Gauge != nil {
+		m.Metric.Gauge.Value = 0
+	}
+	if m.Metric.Histogram != nil {
+		m.Metric.Histogram.SampleCount = 0
+		m.Metric.Histogram.SampleCountFloat = 0
+		m.Metric.Histogram.SampleSum = 0
+		m.Metric.Histogram.Bucket = m.Metric.Histogram.Bucket[:0]
+		m.Metric.Histogram.CreatedTimestamp = nil
+		m.Metric.Histogram.Schema = 0
+		m.Metric.Histogram.ZeroThreshold = 0
+		m.Metric.Histogram.ZeroCount = 0
+		m.Metric.Histogram.ZeroCountFloat = 0
+		m.Metric.Histogram.NegativeSpan = m.Metric.Histogram.NegativeSpan[:0]
+		m.Metric.Histogram.NegativeDelta = m.Metric.Histogram.NegativeDelta[:0]
+		m.Metric.Histogram.NegativeCount = m.Metric.Histogram.NegativeCount[:0]
+		m.Metric.Histogram.PositiveSpan = m.Metric.Histogram.PositiveSpan[:0]
+		m.Metric.Histogram.PositiveDelta = m.Metric.Histogram.PositiveDelta[:0]
+		m.Metric.Histogram.PositiveCount = m.Metric.Histogram.PositiveCount[:0]
+		m.Metric.Histogram.Exemplars = m.Metric.Histogram.Exemplars[:0]
+	}
+	if m.Metric.Summary != nil {
+		m.Metric.Summary.SampleCount = 0
+		m.Metric.Summary.SampleSum = 0
+		m.Metric.Summary.Quantile = m.Metric.Summary.Quantile[:0]
+		m.Metric.Summary.CreatedTimestamp = nil
+	}
+}
+
+func (m *MetricStreamingDecoder) GetMetric() {
+	panic("don't use GetMetric, use Metric directly")
+}
+
+func (m *MetricStreamingDecoder) GetLabel() {
+	panic("don't use GetLabel, use Labels instead")
+}
+
+// Label parses labels into labels scratch builder. Metric name is missing
+// given the protobuf metric model and has to be deduced from the metric family name.
+func (m *MetricStreamingDecoder) Label(b *labels.ScratchBuilder) error {
+	for _, l := range m.labels {
+		if err := parseLabels(m.mData[l.start:l.end], b); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parseLabels is essentially LabelPair.Unmarshal but directly adding into scratch builder
+// and reusing strings.
+func parseLabels(dAtA []byte, b *labels.ScratchBuilder) error {
```

Comment:
At first I was confused due to the name and the fact that it has a for loop inside, but the `b.Add()` bit is outside the loop.

```suggestion
func parseLabel(dAtA []byte, b *labels.ScratchBuilder) error {
```